esmero/strawberryfield
# ISSUE-88: WIP of a JSON API Drush command wrapper

## This commit adds:
See #88 

- A wrapper drush command (example call)
Where the main argument is a pure SBF JSON (for now because it can/should also deal with a FULL ready to be posted JSON API data). It takes a --files argument with a folder of files or a single file (need to check if this is working, this is all quite precarious still), rest are the bundle, the user and pass for jsonapi and a moderation state if people want that to be set.

Right now this actually works and ingests a few files, then the node and connects the dots. But! Since `as:image` etc are still generated by the webform and not as an eventSubscriber, after the creation you still need to go the node, edit via we webform as such. Which is terrible. So next commit will fix that finally!

All this requires also
 - `docker exec -ti esmero-php bash -c "php -dmemory_limit=-1 /usr/bin/composer require drupal/jsonapi_earlyrendering_workaround; drush en jsonapi_earlyrendering_workaround;"`
For some reason (documented in many places like https://www.drupal.org/project/drupal/issues/3072076 there seems to be a render context missing and cache rendering stuff (long expl) is leaking giving us  bad error message at the  end, even when the ingest actually works!

So i added a check, this is a soft dependency since i don't want to depend on that contrib module, but drupal forces me into this situations.

Also @giancarlobi **i need help** with this script. Testing but also making it better. It feels basically just too precarious right now and i need to make it stronger. Will share demo assets for testing after my morning teaching sessions. That is also why i'm making this agains ISSUE-86 and not directly against Beta3. Once tested i can move forward.

This also adds some soft dependencies that should get solved once JSON-API fixes a weird issue with some leaking cache metadata (early rendering, means some code renders something before the response from JSON-API goes back but no caching happens for that). To be honest, why in the hell would you want to cache a POST request to create an entity??

## Steps to test this
- Make sure you have Drush 9 or 10. Not 8, 9 or 10.
- Unzip this folder in archipelago's cloned github repo inside the `d8content` folder
[ados.zip](https://github.com/esmero/strawberryfield/files/4725063/ados.zip)
That will create an `ado` folder with a json file inside an another folder which will contain a poster of trilobites.
- (if on Archipelago-deployment docker, if not you are already an expert and know what to do here1) 
```Shell 
docker exec -ti esmero-php bash -c "php -dmemory_limit=-1 /usr/bin/composer require drupal/jsonapi_earlyrendering_workaround; drush en jsonapi_earlyrendering_workaround; drush cr`
```

- Checkout this branch which also contains the same changes that ISSUE-86 has.
- drush cr again.
- Run the drush command like this (or adapt if not running via docker)
```Shell 
docker exec -ti esmero-php bash -c "drush archipelago:jsonapi-ingest /var/www/html/d8content/ados/ado-photograph.json --bundle=digital_object --uri=http://esmero-web --files=/var/www/html/d8content/ados/ado-photograph-bin --user=jsonapi --password=jsonapi --moderation_state=published"
```
You can also provide an --uuid=someuuid if you want to.  To avoid double ingesting the same one

That should
- upload the single image in that folder
- Take the full expanded JSON, compress, complemente, build an JSON API request, ask for an UUID and create a new Digital Object.

- You will see a LOT of output and that needs to be cleaned up.
- Script is not smart enough yet, so does not do all the checks needed! (HELP)
- Script will upload the same image over and over, but none of them will be attached until the node is processed
-  Check in your site for the content menu, a new published content will appear

@TODO: move the `as:` structure generator out of the webform handler into an event. That way persisting the file and extracting as is done on ingest via JSON-API too.
